### PR TITLE
feat: publish to callback endpoint for tracking

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

File: pkg/udf/rpc/grpc_map.go [hunk 1]
```diff
@@ -141,13 +145,18 @@ func (u *GRPCBasedMap) ApplyMap(ctx context.Context, readMessage *isb.ReadMessag
...
 	writeMessages := make([]*isb.WriteMessage, 0)
-	for _, result := range response.GetResults() {
+	for index, result := range response.GetResults() {
 		keys := result.Keys
 		taggedMessage := &isb.WriteMessage{
 			Message: isb.Message{
 				Header: isb.Header{
 					MessageInfo: parentMessageInfo,
 					Keys:        keys,
+					ID: isb.MessageID{
```
Comment (whynowy) [inline pkg/udf/rpc/grpc_map.go:155]: `ID` was not set?

Comment (yhl25) [reply pkg/udf/rpc/grpc_map.go:155]: it was set inside forward.go

File: pkg/shared/callback/callback.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send request: %w", err)
+	}
+
+	if resp.StatusCode > 299 {
+		return fmt.Errorf("received non-OK response status: %s", resp.Status)
+	}
+
+	_ = resp.Body.Close()
```
Comment (whynowy) [inline pkg/shared/callback/callback.go:243]: Use `defer` before line 239.

File: pkg/shared/callback/callback.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+		opt(dOpts)
+	}
+
+	clientCache, _ := lru.NewWithEvict[string, *http.Client](dOpts.cacheSize, func(key string, value *http.Client) {
+		// Close the client when it's removed from the cache
+		value.CloseIdleConnections()
+	})
+
+	if dOpts.callbackURL != "" {
+		client := &http.Client{
```
Comment (whynowy) [inline pkg/shared/callback/callback.go:41]: This has been existing in `GetClient()`, do not need it here?

File: pkg/sources/source.go [hunk 4]
```diff
@@ -74,7 +75,7 @@ func (sp *SourceProcessor) Start(ctx context.Context) error {
 		healthCheckers           []metrics.HealthChecker
 		idleManager              wmb.IdleManager
 		pipelineName             = sp.VertexInstance.Vertex.Spec.PipelineName
-		vertexName               = sp.VertexInstance.Vertex.Name
+		vertexName               = sp.VertexInstance.Vertex.Spec.Name
```
Comment (whynowy) [inline pkg/sources/source.go:78]: Will existing pipelines get impacted with this fix?

File: pkg/isb/message.go [hunk 5]
```diff
@@ -73,6 +74,23 @@ type Header struct {
...
 
+// MessageID is the message ID of the message which is used for exactly-once-semantics.
+type MessageID struct {
+	// VertexName is the name of the vertex
+	VertexName string
+	// Offset is the offset of the message
+	// NOTE: should be unique across the replicas of the vertex, that is the
+	// reason we don't have a separate replica field in the MessageID
+	Offset string
+	// Index is the index of the message
```
Comment (whynowy) [inline pkg/isb/message.go:85]:
```suggestion
	// Index is the index of a flatmap message, otherwise use 0
```

File: pkg/isb/message.go [hunk 6]
```diff
@@ -73,6 +74,23 @@ type Header struct {
 	Headers map[string]string
 }
 
+// MessageID is the message ID of the message which is used for exactly-once-semantics.
+type MessageID struct {
```
Comment (whynowy) [inline pkg/isb/message.go:78]: Will this change impact existing pipelines?

Comment (yhl25) [reply pkg/isb/message.go:78]: No

File: pkg/sources/source.go (hunk 4 above)
Comment (yhl25) [reply pkg/sources/source.go:78]: No

File: pkg/shared/callback/callback.go (hunk 3 above)
Comment (yhl25) [reply pkg/shared/callback/callback.go:41]: we don't need it

File: pkg/sinks/sink.go [hunk 7]
```diff
@@ -87,7 +92,7 @@ func (u *SinkProcessor) Start(ctx context.Context) error {
 		// create reader for each partition. Each partition is a group in redis
 		for index, bufferPartition := range u.VertexInstance.Vertex.OwnedBuffers() {
 			fromGroup := bufferPartition + "-group"
-			consumer := fmt.Sprintf("%s-%v", u.VertexInstance.Vertex.Name, u.VertexInstance.Replica)
+			consumer := fmt.Sprintf("%s-%v", vertexName, u.VertexInstance.Replica)
```
Comment (whynowy) [inline pkg/sinks/sink.go:95]: This is risky. `vertex.spec.Name` is not unique in one namespace, while `vertex.Name` is.

File: pkg/udf/map_udf.go [hunk 8]
```diff
@@ -125,9 +130,13 @@ func (u *MapUDFProcessor) Start(ctx context.Context) error {
 		return fmt.Errorf("unrecognized isbsvc type %q", u.ISBSvcType)
 	}
 
-	enableMapUdfStream, err := u.VertexInstance.Vertex.MapUdfStreamEnabled()
-	if err != nil {
-		return fmt.Errorf("failed to parse UDF map streaming metadata, %w", err)
+	enableMapStreamStr := os.Getenv(dfv1.EnvMapStreaming)
```
Comment (whynowy) [inline pkg/udf/map_udf.go:133]: Use `pkg/shared/util.LookupEnvBoolOr()`

File: pkg/shared/callback/callback.go [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+// GetClient returns a client for the given URL from the cache
+// If the client is not in the cache, a new one is created.
+func (u *Uploader) GetClient(url string) *http.Client {
+	// Check if the client is in the cache
+	if client, ok := u.clientsCache.Get(url); ok {
+		return client
+	}
+
+	// If the client is not in the cache, create a new one
+	client := &http.Client{
```
Comment (vigith) [inline pkg/shared/callback/callback.go:265]: if the callback server (for this url) is restarted, will the http.Client auto-reconnect internally or will this client become invalid?

**Review by vigith — APPROVED** [review]

**Review by whynowy — APPROVED** [review]

File: examples/1-simple-pipeline.yaml [hunk 10]
```diff
@@ -5,16 +5,20 @@ metadata:
...
   vertices:
     - name: in
+      scale:
+        min: 1
       source:
         # A self data generating source
-        generator:
-          rpu: 5
-          duration: 1s
+        http: {}
```
[the remark this answers is not in the document]
Comment (yhl25) [reply examples/1-simple-pipeline.yaml:12]: This was added by mistake.

File: pkg/shared/callback/callback.go (hunk 9 above)
Comment (yhl25) [reply pkg/shared/callback/callback.go:265]: For the server restart case, If I am not wrong the client doesn't maintain a persistent connection to the server, so it simply creates a new connection when there is a post request, so ideally it should be fine. But still, the client caches the TCP connections. Let me run some tests, If I find any issues I will create a follow-up PR to fix it.

**Pull request merged**